### PR TITLE
feat: strict boolean checks

### PR DESCRIPTION
**Pull request by arty-name**
## fixes KILTProtocol/ticket#2109

These are the changes we will need to have the strict boolean checks. 

The rule has caught a couple of real error cases. Some other requirements are a bit idiosyncratic.

## Checklist:

- [x] I have verified that the code works
- [ ] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [x] I have [left the code in a better state](https://deviq.com/principles/boy-scout-rule)
- [ ] I have documented the changes (where applicable)

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -20,6 +20,7 @@ module.exports = {
   parserOptions: {
     ecmaVersion: 2018,
     sourceType: 'module',
+    project: './tsconfig.build.json'
```
Comment (arty-name) [inline .eslintrc.js:23]: Grants ESLint access to the TS types information

File: packages/core/src/attestation/Attestation.ts [hunk 2]
```diff
@@ -45,7 +45,7 @@ export function verifyDataStructure(input: IAttestation): void {
   }
   DataUtils.validateHash(input.claimHash, 'Claim')
 
-  if (typeof input.delegationId !== 'string' && !input.delegationId === null) {
+  if (typeof input.delegationId !== 'string' && input.delegationId !== null) {
```
Comment (arty-name) [inline packages/core/src/attestation/Attestation.ts:48]: This is a real error

File: packages/core/src/claim/Claim.ts [hunk 3]
```diff
@@ -289,7 +289,8 @@ export function fromCTypeAndClaimContents(
   claimContents: IClaim['contents'],
   claimOwner: DidUri
 ): IClaim {
-  if (ctypeInput.schema) {
+  // TODO: is schema really optional? Yes = fix type. No = use isICType here.
+  if ('schema' in ctypeInput) {
```
Comment (arty-name) [inline packages/core/src/claim/Claim.ts:293]: Need a decision here

File: packages/core/src/credential/Credential.ts [hunk 4]
```diff
@@ -201,13 +201,13 @@ export function verifyDataIntegrity(input: ICredential): boolean {
...
-  const verificationResult = Claim.verifyDisclosedAttributes(input.claim, {
+  const { errors, verified } = Claim.verifyDisclosedAttributes(input.claim, {
     nonces: input.claimNonceMap,
     hashes: input.claimHashes,
   })
   // TODO: how do we want to deal with multiple errors during claim verification?
-  if (!verificationResult.verified)
-    throw verificationResult.errors[0] || new SDKErrors.ClaimUnverifiableError()
+  if (!verified)
+    throw errors.length > 0 ? errors[0] : new SDKErrors.ClaimUnverifiableError()
```
Comment (arty-name) [inline packages/core/src/credential/Credential.ts:210]: There’s no simple way to let TS know that no value exists at some index of array. Empty arrays are also difficult for TS. See [noUncheckedIndexedAccess](https://www.typescriptlang.org/tsconfig#noUncheckedIndexedAccess).

File: packages/core/src/credential/Credential.ts [hunk 5]
```diff
@@ -255,11 +255,11 @@ export function verifyDataStructure(input: ICredential): void {
     throw new SDKErrors.ClaimNonceMapMalformedError()
   }
 
-  if (!input.claimHashes) {
+  if (!('claimHashes' in input)) {
     throw new SDKErrors.DataStructureError('claim hashes not provided')
   }
 
-  if (typeof input.delegationId !== 'string' && !input.delegationId === null) {
+  if (typeof input.delegationId !== 'string' && input.delegationId !== null) {
```
Comment (arty-name) [inline packages/core/src/credential/Credential.ts:262]: Again the same error

File: packages/core/src/delegation/DelegationNode.spec.ts [hunk 6]
```diff
@@ -26,9 +26,9 @@ let nodes: Record<string, DelegationNode> = {}
 
 jest.mock('./DelegationNode.chain', () => ({
   getChildren: jest.fn(async (node: DelegationNode) =>
-    node.childrenIds.map((id) => nodes[id] || null)
+    node.childrenIds.map((id) => (id in nodes ? nodes[id] : null))
```
Comment (arty-name) [inline packages/core/src/delegation/DelegationNode.spec.ts:29]: Here `nodes` is a Record of string to object, so TS does not expect `nodes[id]` to be undefined. Modifying the type of `nodes` to be nullable results in a ton of unchecked access errors in this test, so I opted for the smaller set of fixes.

File: packages/did/src/DidDetails/FullDidDetails.ts [hunk 7]
```diff
@@ -130,7 +130,7 @@ export async function authorizeExtrinsic(
   }
 
   const [signingKey] = getKeysForExtrinsic(did, extrinsic)
-  if (!signingKey) {
+  if (signingKey === undefined) {
```
Comment (arty-name) [inline packages/did/src/DidDetails/FullDidDetails.ts:133]: Again no notion of empty arrays. I think we should rather avoid them.

File: packages/vc-export/src/vc-js/suites/KiltAbstractSuite.ts [hunk 8]
```diff
@@ -83,7 +83,7 @@ export abstract class KiltAbstractSuite extends suites.LinkedDataProof {
   }): Promise<boolean> {
     const { proof } = options
     const compact = await this.compactProof(proof, options)
-    const type = compact['@type'] || compact.type
+    const type = compact['@type'] ?? compact.type
```
Comment (arty-name) [inline packages/vc-export/src/vc-js/suites/KiltAbstractSuite.ts:86]: In all cases of amorphous JSON we get a type which can be a numeric value of `0`, and should use nullish coalescence

File: packages/vc-export/src/verificationUtils.ts [hunk 9]
```diff
@@ -86,18 +86,19 @@ export async function verifySelfSignedProof(
...
     }
-    const dereferenced = documentLoader
-      ? await documentLoader(verificationMethod)
-      : undefined
-    if (!dereferenced?.document) {
+    if (typeof documentLoader !== 'function') {
+      throw new Error('did you select an appropriate document loader?')
+    }
+    const dereferenced = await documentLoader(verificationMethod)
+    if (!('document' in dereferenced)) {
```
Comment (arty-name) [inline packages/vc-export/src/verificationUtils.ts:93]: This whole block is weird

File: packages/chain-helpers/src/blockchain/SubscriptionPromise.ts [hunk 10]
```diff
@@ -39,13 +39,14 @@ export function makeSubscriptionPromise<SubscriptionType>(
   const subscription: (value: SubscriptionType) => void =
     typeof rejectOn === 'function'
       ? (value) => {
-          if (rejectOn(value)) reject(value)
-          if (resolveOn(value)) resolve(value)
+          // eslint-disable-next-line no-extra-boolean-cast
```
Comment (arty-name) [inline packages/chain-helpers/src/blockchain/SubscriptionPromise.ts:42]: Should we disable this rule globally?

File: .eslintrc.js (hunk 1 above)
Comment (tjwelde) [reply .eslintrc.js:23]: otherwise it doesn't work?

Comment (arty-name) [reply .eslintrc.js:23]: Yeah, the rule doesn’t even attempt to work without it, it’s a hard requirement.

File: packages/chain-helpers/src/blockchain/SubscriptionPromise.ts (hunk 10 above)
Comment (arty-name) [reply packages/chain-helpers/src/blockchain/SubscriptionPromise.ts:42]: Found out that the reason is our poor type

File: packages/chain-helpers/src/blockchain/Blockchain.ts [hunk 11]
```diff
@@ -139,11 +139,9 @@ export async function submitSignedTx(
 
   function handleDisconnect(): void {
     const result = new SubmittableResult({
-      events: latestResult.events || [],
+      events: latestResult.events,
```
Comment (tjwelde) [inline packages/chain-helpers/src/blockchain/Blockchain.ts:142]: That change looks not correct to me. `latestResult` can be declared, but not instantiated, before reaching `handleDisconnect`. Same with the status line below.

File: packages/core/src/claim/Claim.ts (hunk 3 above)
Comment (tjwelde) [reply packages/core/src/claim/Claim.ts:286]: The type definition of `ICType` does not define schema to be optional and that is good and correct.
Thinking about JS people though, it is a good idea to use our typeguards to protect against mistakes though.

File: packages/core/src/credential/Credential.ts [hunk 12]
```diff
@@ -226,19 +226,19 @@ export function verifyDataIntegrity(input: ICredential): boolean {
...
   if (!input.claim.owner) {
     throw new SDKErrors.OwnerMissingError()
   }
-  if (!input.legitimations && !Array.isArray(input.legitimations)) {
+  if (!Array.isArray(input.legitimations)) {
     throw new SDKErrors.LegitimationsMissingError()
   }
 
-  if (!input.claimNonceMap) {
+  if (!('claimNonceMap' in input)) {
```
Comment (tjwelde) [inline packages/core/src/credential/Credential.ts:241]: Not sure about all these `in` checks. `in` returns true, if the property is there, but set to undefined, which is possible in some of our types.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/in#using_in_with_deleted_or_undefined_properties

What are your thoughts on this?

File: packages/vc-export/src/verificationUtils.ts (hunk 9 above)
Comment (tjwelde) [reply packages/vc-export/src/verificationUtils.ts:93]: there are a LOT of conditionals

File: packages/did/src/DidDetails/FullDidDetails.ts (hunk 7 above)
Comment (tjwelde) [reply packages/did/src/DidDetails/FullDidDetails.ts:133]: don't really understand what you mean

Comment (tjwelde) [reply packages/did/src/DidDetails/FullDidDetails.ts:133]: we clarified it. TS is not complaining, that `signingKey` could be undefined.

File: .eslintrc.js [hunk 13]
```diff
@@ -73,6 +74,11 @@ module.exports = {
       },
     ],
     '@typescript-eslint/ban-ts-comment': 'warn',
+    '@typescript-eslint/strict-boolean-expressions': ['error', {
+      allowNumber: false,
```
[the remark this answers is not in the document]
Comment (tjwelde) [reply .eslintrc.js:78]: it is

File: packages/core/src/credential/Credential.ts (hunk 12 above)
Comment (tjwelde) [reply packages/core/src/credential/Credential.ts:241]: Function expects ICredential, not e.g. Partial<ICredential>, so it is technically correct to ignore the potential undefined properties. There is a bigger question on if we should change the verify functions to rather be type guards, or at least expect an input with potential undefined properties.

File: .eslintrc.js [hunk 14]
```diff
@@ -73,6 +74,11 @@ module.exports = {
...
+      allowNullableString: true,
```
[the remark this answers is not in the document]
Comment (tjwelde) [reply .eslintrc.js:79]: maybe we tackle that later

File: packages/core/src/claim/Claim.ts (hunk 3 above)
Comment (arty-name) [reply packages/core/src/claim/Claim.ts]: Added the type guard in https://github.com/KILTprotocol/sdk-js/pull/612/commits/e2abc8dc05b0343ea70e1959aa2a21578d445271

**Review by tjwelde — APPROVED** [review]
LGTM

**Pull request merged**